plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTab.java

```diff
@@ -352,11 +346,14 @@ void updateRunnerButtons(MajorVersion majorVersion) {
   private void updateHierarchy(final MajorVersion majorVersion)
       throws InvocationTargetException, InterruptedException {
     // blocking call (regardless of "fork"), returning only after the inner runnable completes
-    getLaunchConfigurationDialog().run(true, true, new IRunnableWithProgress() {
+    getLaunchConfigurationDialog().run(true /*fork*/, true /*cancelable*/,
+        new IRunnableWithProgress() {
       @Override
       public void run(IProgressMonitor monitor)
           throws InvocationTargetException, InterruptedException {
-        hierarchy = getPipelineOptionsHierarchy(majorVersion, monitor);
+        SubMonitor subMonitor = SubMonitor.convert(
```

Comment:
I believe you need to call done() on the monitor at the end here.  That's the protocol on IProgressMonitor (`beginTask()` must have a corresponding `done()`), and the same is said on `SubMonitor#convert()`

> It is not necessary to call done() on the result, but the caller is responsible for calling done() on the argument. Calls beginTask on the argument.

Though the examples of use on `SubMonitor` are confusing.

Reply:
We used `SubMonitor` everywhere, but we almost never called `done()`.  This is quite new to me. Maybe we should file an issue.

Reply:
I was reading comments of `SubMonitor`:

```
 * <p><b>BEST PRACTISES:</b></p>
 * 
 * <p>We recommend that newly-written methods follow the given contract:</p>
 * <ul>
 * <li>It WILL call beginTask on the IProgressMonitor.</li>
 * <li>It WILL accept a null argument, indicating that no progress should be reported and the operation cannot be cancelled.</li>
 * <li>It WILL NOT call done on the IProgressMonitor, leaving this responsibility up to the caller.</li>
 * </ul>
 * <p>If you wish to follow these conventions, you may copy and paste the following text into your method's JavaDoc:</p>
 * 
 * <pre>@param monitor the progress monitor to use for reporting progress to the user. It is the caller's responsibility
 *        to call done() on the given monitor. Accepts <code>null</code>, indicating that no progress should be
 *        reported and that the operation cannot be cancelled.</pre>
```

Note

```
you may copy and paste the following text into your method's JavaDoc:</p>
```

Note that following Javadoc of `convert()` is documentation for callers of `convert()`. That means, the caller in the Javadoc refers to the caller of the method that uses `convert()`. (In our case as an example, is the caller of `run()`.)

> It is not necessary to call done() on the result, but the caller is responsible for calling done() on the argument. Calls beginTask on the argument.

Reading this again, it all makes sense. We shouldn't call `done()` as the Javadoc says. It is the responsibility of the caller of our method `run()` to call `done()`.